File: src/process-command.ts

import {
  Command
} from './commands/commands';

import {
  getAllKeys,
  index,
  isIndexed,
  isDependedOn,
  remove
} from './data-access/data-access';

import {
  INDEX_NUMBER,
  REMOVE_NUMBER,
  QUERY_NUMBER
} from './utils/constants';

import {
  reduceBools
} from './utils/helpers';

export function processCommand(command: Command): Promise<boolean> {
  if (command.type === INDEX_NUMBER) {
    return processIndexCommand(command);
  } else if (command.type === REMOVE_NUMBER) {
    return processRemoveCommand(command);
  } else if (command.type === QUERY_NUMBER) {
    return processQueryCommand(command);
  }
  throw new Error('Unknown Command Type');
}

export function processIndexCommand(command: Command): Promise<boolean> {
  // okay, first thing we need to check is if all of the dependencies are indexed
  const promises = command.dependencies.map(dependency => isIndexed(dependency));
  return Promise.all(promises).then((results: boolean[]) => {
    const allSucceeded = results.length ? reduceBools(results) : true;
    if (allSucceeded) {
      // sweet, we can go ahead and index it since all of the dependencies are good to go
      return index(command.packageName, command.dependencies).then(() => {
        return true;
      })
    }
    return Promise.resolve(false);
  });
}

export function processRemoveCommand(command: Command): Promise<boolean> {
  // okay, first thing we need to do is check if this command is a dependency of another command.
  // if it is, we cannot remove it.
  return isDependedOn(command.packageName).then((dependedOn) => {
    
    if (dependedOn) {
      return false;
    }

    // it's not depended on, so we can remove it
    
    return remove(command.packageName).then(() => {
      return true;
    });
  });
}

export function processQueryCommand(command: Command): Promise<boolean> {
  return isIndexed(command.packageName);
}

File: go/process-command.go
package main

import (
	"errors"
)

func processCommand(command *Command) (bool, error) {
	if command.Type == INDEX_NUMBER {
		return processIndexCommand(command)
	} else if command.Type == REMOVE_NUMBER {
		return processRemoveCommand(command)
	} else if command.Type == QUERY_NUMBER {
		return processQueryCommand(command)
	}
	return false, errors.New("Unknown Command Type")
}

func processIndexCommand(command *Command) (bool, error) {
	Debugf("Starting to process index command for %s", command.PackageName)
	// okay, first thing we need to do is to check if all of the dependencies are indexed
	for i := range command.Dependencies {
		dependency := command.Dependencies[i]
		// if it's not indexed, then return false
		if !IsIndexed(dependency) {
			Debugf("Failed to process index command for %s due to dependency %s not being indexed", command.PackageName, dependency)
			return false, nil
		}
	}

	IndexPackage(command.PackageName, command.Dependencies)
	// okay cool, we've checked each dependency and they all appear to be indexed, so
	// we're good to go
	Debugf("Successfully processed index command for %s", command.PackageName)
	return true, nil
}

func processRemoveCommand(command *Command) (bool, error) {
	Debugf("Starting to process remove command for %s", command.PackageName)
	// okay, first thing we need to do is check if this command is a dependency of another command.
	// if it is, we cannot remove it.
	if IsDependedOn(command.PackageName) {
		// it's depended on, so it can't be removed
		Debugf("Failed to process remove command for %s since it's depended on by another package", command.PackageName)
		return false, nil
	}

	// w00t, we can remove it
	RemovePackage(command.PackageName)
	Debugf("Successfully processed remove command for %s", command.PackageName)
	return true, nil
}

func processQueryCommand(command *Command) (bool, error) {
	Debugf("Starting to process query command for %s", command.PackageName)
	result := IsIndexed(command.PackageName)
	Debugf("Successfully processed query command for %s, result is %t", command.PackageName, result)
	return result, nil
}
